CNV-22037: new content volume snapshot (#65659)
openshift/openshift-docs: modules/virt-enabling-volume-snapshot-boot-source.adoc

// Module included in the following assembly:
//
// * virt/storage/virt-automatic-bootsource-updates.adoc
//

:_content-type: PROCEDURE
[id="virt-enabling-volume-snapshot-boot-source_{context}"]
= Enabling volume snapshot boot sources

> Enable volume snapshot boot sources by setting the parameter in the `DataImportCron` namespace. Although `DataImportCron` was originally designed to maintain only PVC sources, volume snapshot sources scale better than PVC sources for certain storage types.

Reviewer: `Enable volume snapshot boot sources by setting the parameter in the DataImportCron namespace.` - not sure I understand... I would suggest removing this. 
@akalenyu, could you please look, maybe I'm missing the point here?

> Enable volume snapshot boot sources by setting the parameter in the `DataImportCron` namespace. Although `DataImportCron` was originally designed to maintain only PVC sources, volume snapshot sources scale better than PVC sources for certain storage types.

Reviewer 2: The above is confusing.  We set the parameter in the StorageProfile associated with the storage class that is being used to store operating system base images.

> Enable volume snapshot boot sources by setting the parameter in the `StorageProfile` associated with the storage class that stores operating system base images. Although `StorageProfile` was originally designed to maintain only PVC sources, volume snapshot sources scale better than PVC sources for certain storage types.

Reviewer: ```suggestion
Enable volume snapshot boot sources by setting the parameter in the `StorageProfile` associated with the storage class that stores operating system base images. Although `DataImportCron` was originally designed to maintain only PVC sources, `VolumeSnapshot` sources scale better than PVC sources for certain storage types.
```

[NOTE]
====
Use volume snapshots on a storage profile that is proven to scale better when cloning from a single snapshot.
====

.Prerequisites

* You must have access to a volume snapshot with the operating system image.

> * The storage profile must support snapshotting.

Reviewer: ```suggestion
* The storage must support snapshotting.
```

> * The storage profile must support snapshotting.

Reviewer 2: The storage driver must support CSI snapshots.

Author: @aglitke @jpeimer I updated this to `The storage must support snapshotting`, is this enough or should I include "CSI"?.

Reviewer: This should be enough, because k8s doc says `VolumeSnapshot support is only available for CSI drivers.`
https://kubernetes.io/docs/concepts/storage/volume-snapshots/

.Procedure

. Open the storage profile object that corresponds to the storage class used to provision boot sources by running the following command:
+
[source,terminal,subs="attributes+"]
----

> $ oc edit storageprofile <storage_class> -oyaml

Reviewer: So here we don't need -oyaml in the end 
```suggestion
$ oc edit storageprofile <storage_class>
```

Reviewer: I don't understand. It wasn't fixed, but it was marked resolved.

----

. Review the `dataImportCronSourceFormat` specification of the `StorageProfile` to confirm whether or not the VM is using PVC or volume snapshot by default.

. Edit the storage profile, if needed, by updating the `dataImportCronSourceFormat` specification to `snapshot`.
+
.Example storage profile
[source,yaml]
----
apiVersion: cdi.kubevirt.io/v1beta1
kind: StorageProfile
metadata:
# ...
spec:
  dataImportCronSourceFormat: snapshot
----

.Verification

. Open the storage profile object that corresponds to the storage class used to provision boot sources.
+
[source,terminal,subs="attributes+"]
----

> $ oc edit storageprofile <storage_class>

Reviewer: ```suggestion
$ oc get storageprofile <storage_class> -oyaml
```

Reviewer 3: Why?

Reviewer: I don't think we need to `edit` the storageprofile to verify that changes were applied. It's enough to just look at it.

Reviewer 3: right this is the verification part. yep

> $ oc get storageprofile <storage_class>

Reviewer: ```suggestion
$ oc get storageprofile <storage_class> -oyaml
```

> $ oc get storageprofile <storage_class>

Reviewer: But here we need -oyaml in the end
```suggestion
$ oc get storageprofile <storage_class> -oyaml
```

Reviewer: And this comment

----

> . Confirm that the `dataImportCronSourceFormat` specification of the `StorageProfile` is set to 'snapshot', and that any data source objects that the `DataImportCron` points to now reference volume snapshopts.

Reviewer: ```suggestion
. Confirm that the `dataImportCronSourceFormat` specification of the `StorageProfile` is set to 'snapshot', and that any `DataSource` objects that the `DataImportCron` points to now reference volume snapshopts.
```

> . Confirm that the `dataImportCronSourceFormat` specification of the `StorageProfile` is set to 'snapshot', and that any `DataSource` objects that the `DataImportCron` points to now reference volume snapshopts.

Reviewer 4: ```suggestion
. Confirm that the `dataImportCronSourceFormat` specification of the `StorageProfile` is set to 'snapshot', and that any `DataSource` objects that the `DataImportCron` points to now reference volume snapshots.
```

> You can now use this boot source to create {op-system-base} virtual machines.

Reviewer: ```suggestion
You can now use these boot sources to create virtual machines.
```
(because there're not only RHEL images, but also centos, fedora etc.)

// Module included in the following assembly:
//
// * virt/storage/virt-automatic-bootsource-updates.adoc
//

:_content-type: PROCEDURE
[id="virt-enabling-volume-snapshot-boot-source_{context}"]
= Enabling volume snapshot boot sources

> Enable volume snapshot boot sources by setting the parameter in the `DataImportCron` namespace. Although `DataImportCron` was originally designed to maintain only PVC sources, volume snapshot sources scale better than PVC sources for certain storage types.

Reviewer 3: by setting the parameter in the `StorageProfile` (which is not namespaced)

[NOTE]
====

> It is recommended to use volume snapshots on a storage profile that is proven to scale better when cloning from a single snapshot.

Reviewer 5: We're typically supposed to [direct users to take a recommended action rather than stating something is recommended](https://redhat-documentation.github.io/supplementary-style-guide/#recommend), though up to you if you feel strongly about keeping it:

```suggestion
Use volume snapshots on a storage profile that is proven to scale better when cloning from a single snapshot.
```

====

.Prerequisites

> * You must have access to a volume snapshot with the operating system image.

Reviewer 3: So this can remain prerequisite remains `You must have access to a web page with the operating system image`;
This is because our product takes care of converting the imported source to a VolumeSnapshot.

Maybe a couple of other prerequisites are:
- The storage supports snapshotting
- Storage provider proven to scale better cloning from a single snapshot (not a must, I guess if someone is reading this, they are convinced this is the case)

* The storage profile must support snapshotting.

.Procedure

> . Open the storage profile object that corresponds to the storage class used to provision boot sources.

Reviewer 5: ```suggestion
. Open the storage profile object that corresponds to the storage class used to provision boot sources by running the following command:
```

+
[source,terminal,subs="attributes+"]
----
$ oc edit storageprofile <storage_class>
----

> . Review the `dataImportCronSourceFormat` specification of the `DataImportCron` namespace to confirm whether or not the VM is using PVC or volume snapshot by default.

Reviewer 3: `dataImportCronSourceFormat` is a specification under the StorageProfile object, not DataImportCron

. Edit the storage profile, if needed, by updating the `dataImportCronSourceFormat` specification to `snapshot`.
+
.Example storage profile
[source,yaml]
----
apiVersion: cdi.kubevirt.io/v1beta1
kind: StorageProfile
metadata:
# ...
spec:
  dataImportCronSourceFormat: snapshot
----

.Verification

. Open the storage profile object that corresponds to the storage class used to provision boot sources.

> 

Reviewer 5: Seems to render fine but you could delete this blank like if you wanted to.

+
[source,terminal,subs="attributes+"]
----
$ oc edit storageprofile <storage_class>
----

> . Confirm that the `dataImportCronSourceFormat` specification of the `DataImportCron` namespace is using volume snapshots by default.

Reviewer 3: I think verification that this "went through" would be to:
- Ensure the storage profile status section reflects the change (basically that `storageprofile.status.dataImportCronSourceFormat` is equal to "snapshot" now)
- Inspect the DataSource objects which the DataImportCron points to, to ensure they now reference VolumeSnapshots instead of PVCs. This may take a little longer

Also, dataImportCronSourceFormat is a specification under the StorageProfile object, not DataImportCron
so the namespace word throws me off. Am I missing something? StorageProfile is cluster wide
